[Core] Remove lora additional vocabulary

ahengljh opened the pull request:
## Purpose

  This PR removes the ability for each LoRA adapter to define its own additional
  vocabulary, as proposed in issue #23474. All LoRA adapters for a given model now
  share the same vocabulary as the base model.

  ## Test Plan

  1. Test basic LoRA initialization:
  ```python
  import vllm
  llm = vllm.LLM(model="Qwen3-8B", enable_lora=True)
```

  2. Test LoRA adapter loading (with existing adapters):
  ```python
  from vllm.lora.request import LoRARequest
  lora_request = LoRARequest("adapter1", 1, "/path/to/lora/adapter")
  outputs = llm.generate(["Test prompt"], lora_request=lora_request)
```

  4. Run LoRA-related tests:
  ```python
  pytest tests/lora/test_lora_manager.py -v
  pytest tests/lora/test_layers.py -v 
```

  ## Test Result

  - ✅ Successfully initialized vLLM with enable_lora=True without AttributeError
  - ✅ LoRAConfig no longer has lora_extra_vocab_size attribute
  - ✅ All model files updated to remove additional vocabulary references


  (Optional) Documentation Update

  No documentation updates required as this removes an undocumented internal feature.
  <details>
  <summary> Essential Elements of an Effective PR Description Checklist </summary>
  - [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR
  will resolve)".
  - [X] The test plan, such as providing test command.
  - [X] The test results, such as pasting the results comparison before and after, or
  e2e results
  - [ ] (Optional) The necessary documentation update, such as updating
  `supported_models.md` and `examples` for a new model.
 </details>
  BEFORE SUBMITTING, PLEASE READ https://docs.vllm.ai/en/latest/contributing

ahengljh commented:
Allow me to fix some naive mistake here

ahengljh commented:
> As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`

Oh I see. So we should hold this PR and just add warning info to users at this moment right?

jeejeelee commented:
> > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> 
> Oh I see. So we should hold this PR and just add warning info to users at this moment right?

I think so

ahengljh commented:
> > > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> > 
> > 
> > Oh I see. So we should hold this PR and just add warning info to users at this moment right?
> 
> I think so

I just made another one-line PR #23635 for adding the warning, and we may hold this PR until you believe it's time to deprecate this feature completely. What's your opinion on this?

ahengljh commented:
> > > > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> > > 
> > > 
> > > Oh I see. So we should hold this PR and just add warning info to users at this moment right?
> > 
> > 
> > I think so
> 
> I just made another one-line PR #23635 for adding the warning, and we may hold this PR until you believe it's time to deprecate this feature completely. What's your opinion on this?

The #23635 for add warning message is merged. Now we may talk about how about changes in this PR? If we would deprecate this feature completely in v0.12.0, what else should be ready adding to this PR? So that I can make it and waiting for the time.

zhuohan123 commented:
@ahengljh Any blockers for us to go ahead, pass CI, and merge this PR? Happy to help on any blockers to merge the PR.

ahengljh committed:
Remove LoRA additional vocabulary support

This commit completely removes the deprecated lora_extra_vocab_size feature
that allowed LoRA adapters to extend model vocabulary. The feature has been
deprecated and is no longer needed.

Changes:
- Remove lora_extra_vocab_size field from LoRAConfig
- Remove extra_vocab_size parameter from all punica wrapper methods
- Update embeddings_indices calculation to use only base vocab size
- Remove all lora_vocab = 0 assignments from model files
- Fix IndentationError in models caused by empty if blocks
- Update all tests to use new function signatures
- Remove all references and comments about extra vocabulary support

The removal simplifies the codebase and eliminates dead code paths that
were no longer being used.

Signed-off-by: Jinheng Li <ahengljh@gmail.com>

zhuohan123 commented:
@ahengljh We are ready to merge the code now!

jeejeelee commented:
@ahengljh  If you are not available, I'll continue pushing forward with this PR tomorrow

ahengljh commented:
> @ahengljh If you are not available, I'll continue pushing forward with this PR tomorrow

I will continue this, don’t worry

ahengljh committed:
pre-commit changes

Signed-off-by: Jinheng LI <jinhengli@utexas.edu>

ahengljh committed:
Fix unused variable errors after LoRA vocabulary removal

Remove unused lora_config assignments in model classes where the LoRA
additional vocabulary support was removed. The lora_config variable
was being assigned but never used after the vocabulary size logic
was simplified.

Signed-off-by: Jinheng Li <ahengljh@gmail.com>

ahengljh committed:
Remove lora_extra_vocab_size from LoRA config directly

Signed-off-by: Jinheng Li <ahengljh@gmail.com>

ahengljh commented on vllm/engine/arg_utils.py:
```diff
@@ -398,7 +398,6 @@ class EngineArgs:
     fully_sharded_loras: bool = LoRAConfig.fully_sharded_loras
     max_cpu_loras: Optional[int] = LoRAConfig.max_cpu_loras
     lora_dtype: Optional[Union[str, torch.dtype]] = LoRAConfig.lora_dtype
-    lora_extra_vocab_size: int = LoRAConfig.lora_extra_vocab_size
```
Done, but I'm not quite sure about changes in vllm/lora/layers/logits_processor.py and vllm/lora/layers/vocal_parallel_embedding.py, please help review

zhuohan123 commented:
@ahengljh Just FYI #24078 is merged so we don't have tokenizer group in vLLM anymore.

hmellor commented:
@jeejeelee could you give this a review to ensure LoRA will still behave as expected with the additional vocab feature removed?

ahengljh commented on tests/lora/test_layers.py:
```diff
@@ -293,8 +293,7 @@ def create_random_embedding_layer():
         punica_wrapper.update_metadata(lora_mapping, id_to_index, max_loras,
-                                       vocab_size,
-                                       lora_config.lora_extra_vocab_size)
+                                       vocab_size)
```
Following the removal of LoRA additional vocabulary support, the following
   tests in tests/lora/test_layers.py are now failing:

  - test_embeddings_with_new_embeddings - Tests embedding expansion with
  additional vocabulary tokens
  - test_lm_head_logits_processor - Tests logits processing with extended
  vocabulary

  These tests specifically validate the removed functionality where LoRA
  adapters could introduce new tokens beyond the base model's vocabulary
  through embeddings_tensor.

I am not quite sure about these

jeejeelee replied:
Okay, I will check it out ASAP

jeejeelee commented:
I have trigged the lora testing to verify these changes

jeejeelee commented:
@ahengljh  could we speed up a bit? Thank you.

ahengljh committed:
Merge remote-tracking branch 'origin/main' into remove-lora-additional-vocabulary
